### PR TITLE
feat(Turborepo): Add spaces_id to layered config

### PR DESCRIPTION
**Pull request by gsoltis**
### Description

The core goal of this PR is to make `run.run` immutable. This enforces that our config and options are all resolved before we start a run. The last remaining piece of that was picking up the `spaces_id` from the layered config system. 

Adding `spaces_id` to the layered config system necessitated a bunch of other changes due to the fact that it is not contained within the `remoteCache` field in `turbo.json`. The config system previously assumed that any config file could contain the entirety of the config, which wasn't strictly true, but now is even less true. 

To that end, I created a separate struct to represent just the `remoteCache` field in `turbo.json`. It shares much, but not all of the parser for a config file. I split the rendering of `ConfigurationOptions` into two structs, since we want to render them differently in different circumstances (`turbo info` vs `turbo.json` synthesis in `turbo prune`). This had the upshot of letting us skip rendering a) empty fields in pruned repos and b) a non-functional `token` field in pruned repos.

### Testing Instructions

Updated integration tests for expected renderings of the config. Added unit tests for spaces_id.
Deleted the test for not parsing a token from the `remoteCache` field in `turbo.json`, the type system no longer allows us to parse a token since it's no longer directly a `ConfigurationOptions` instance.

Closes TURBO-2313

### CODE REVIEW

File: crates/turborepo-lib/src/config.rs [hunk 1]
```diff
@@ -211,11 +215,15 @@ impl ConfigurationOptions {
...
+
+    pub fn spaces_id(&self) -> Option<&str> {
+        self.spaces_id.as_deref()
+    }
 }
 
 // Maps Some("") to None to emulate how Go handles empty strings
 fn non_empty_str(s: Option<&str>) -> Option<&str> {
-    s.and_then(|s| (!s.is_empty()).then_some(s))
+    s.filter(|s| !s.is_empty())
```
Comment (chris-olszewski) [inline crates/turborepo-lib/src/config.rs:226]: TIL about `Option::filter`, nice!

File: crates/turborepo-lib/src/turbo_json/parser.rs [hunk 2]
```diff
@@ -320,6 +321,100 @@ impl Deserializable for ConfigurationOptions {
...
+    fn deserialize(
+        value: &impl DeserializableValue,
+        name: &str,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self> {
+        value.deserialize(RawRemoteCacheOptionsVisitor, name, diagnostics)
+    }
+}
+
+struct RawRemoteCacheOptionsVisitor;
```
Comment (chris-olszewski) [inline crates/turborepo-lib/src/turbo_json/parser.rs:334]: With this I think we can get rid of the `impl Deserializable for ConfigurationOptions` along with the `ConfigurationOptionsVisitor`

File: crates/turborepo-lib/src/turbo_json/parser.rs [hunk 3]
```diff
@@ -320,6 +321,100 @@ impl Deserializable for ConfigurationOptions {
...
+        let mut result = RawRemoteCacheOptions::default();
+        for (key, value) in members.flatten() {
+            // Try to deserialize the key as a string.
+            // We use `Text` to avoid an heap-allocation.
+            let Some(key_text) = Text::deserialize(&key, "", diagnostics) else {
+                // If this failed, then pass to the next key-value pair.
+                continue;
+            };
+            match key_text.text() {
+                "apiUrl" => {
```
Comment (chris-olszewski) [inline crates/turborepo-lib/src/turbo_json/parser.rs:359]: I think this is fine, but just flagging that this is a breaking change from previous behavior we inherited from Go with all the weird casings for these keys.

**Review by chris-olszewski — APPROVED** [review]

File: crates/turborepo-lib/src/turbo_json/parser.rs (hunk 2 above)
Comment (gsoltis) [reply crates/turborepo-lib/src/turbo_json/parser.rs:334]: I don't think we can. We also deserialize other files via that visitor. The change I made is specific to the `remoteCache` key in `turbo.json`, all other parsing is left unchanged.

File: crates/turborepo-lib/src/turbo_json/parser.rs (hunk 3 above)
Comment (gsoltis) [reply crates/turborepo-lib/src/turbo_json/parser.rs:359]: It's a change only for the `remoteCache` key in `turbo.json`. All of the other config files still go through the other parser and support all of the variants.

**Pull request merged**